### PR TITLE
This took forever, but leaving the debugging challenge was the right …

### PR DESCRIPTION
**Pull request by drfudd2001**
…way to go

# Submitting Week 3 Assessment

### ✅ Completed Files

Mark the following as confirmation that the files in this repository are complete.

- ✅ Coding Challenges
- ✅ Interview Questions
- ✅ React Debugging

### 📝 Feedback

If you would like specific feedback on a particular problem please leave details here:

"Work quality is lacking, wife and daughter bot sick and needing attention. 🤧 😩"

### CODE REVIEW

File: code-challenges.test.js [hunk 1]
```diff
@@ -12,30 +12,152 @@
...
+    // The length of the result should be equal to n
+    expect(result.length).toEqual(number)
+    // The first two numbers in the result should be 1
+    expect(result[0]).toEqual(1)
+    expect(result[1]).toEqual(1)
+    // Each subsequent number in the result should be the sum of the two preceding numbers
+    for (let i = 2; i < number; i++) {
+      expect(result[i]).toEqual(result[i - 1] + result[i - 2])
+      }
+  })
```
Comment (SunkissedQueen) [inline code-challenges.test.js:85]: Though this works, the structure can be cleaned up some. Checkout this article https://webtips.dev/webtips/jest/jest-best-practices

The test function is meant to test separate behaviors of your function. Since we are only testing one function (one behavior) then you will only need one test.

Each expect statement will take care of the edge cases of the values that the function call takes in.

File: code-challenges.test.js [hunk 2]
```diff
@@ -12,30 +12,152 @@
...
+    // Calculate the next number in the sequence by adding the two previous numbers
+    const nextNumber = sequence[i - 1] + sequence[i - 2]
+
+    // Add the next number to the sequence array
+      sequence.push(nextNumber)
+  }
+
+  // Return the generated sequence array
+  return sequence
+}
```
Comment (SunkissedQueen) [inline code-challenges.test.js:131]: nice work taking care of all the edge cases

File: code-challenges.test.js [hunk 3]
```diff
@@ -46,13 +168,93 @@ const studyMinutesWeek2 = {
...
+
+  // Create an array called valuesArr that contains the values of the studyMinutes object
+  const valuesArr = Object.values(studyMinutes)
+
+  // Use the sort method to sort valuesArr in ascending order, using a callback function to compare values
+  valuesArr.sort((a, b) => a - b)
+
+  // Return the sorted array valuesArr
+  return valuesArr
+}
```
Comment (SunkissedQueen) [inline code-challenges.test.js:235]: I love all the tools that can be used in the programming language. Wonderful use of those tools

File: code-challenges.test.js [hunk 4]
```diff
@@ -64,4 +266,129 @@ const accountTransactions2 = [250, -89, 100, -96]
...
+  // Use the map() method to loop through the array and accumulate the sum
+  return arr.map((val) => {
+
+    // For each value in the array, add it to the sum and update the 'sum' variable
+    sum += val
+
+    // Return the updated 'sum' value as the element in the new array
+    return sum
+  })
+}
```
Comment (SunkissedQueen) [inline code-challenges.test.js:339]: I can see that you are very comfortable on the coding challenges

File: interview-questions.md [hunk 5]
```diff
@@ -10,46 +10,142 @@ Then, research the question to expand on your answer. Even if you feel you have
...
+
 Researched answer:
 
-1. What is yarn? What file(s) are modified when you run the command yarn in your terminal?
+- JSX is a syntax extension for JavaScript that allows you to write HTML-like code within your JavaScript code, primarily used in React applications to describe the structure and content of components. It lets you combine HTML and JavaScript into a single file, making your code more modular and easier to manage. The main syntax difference between HTML and JSX is the way attributes are defined. In HTML, attributes are defined using the regular HTML syntax, where you have the attribute name followed by an equals sign, then the value in quotes. For example: `<img src="image.jpg" alt="An image">`. In JSX, attributes are defined using a JavaScript-like syntax, where you use curly braces to insert JavaScript expressions. Also, JSX uses className instead of class to define a CSS class for an element, because class is a reserved keyword in JavaScript. For example: `<img src={imageUrl} alt="An image" />`. Additionally, in JSX, you can write HTML-like syntax directly within your JavaScript code, which allows you to create reusable components and keep your code organized.
```
Comment (SunkissedQueen) [inline interview-questions.md:17]: great content here

File: interview-questions.md [hunk 6]
```diff
@@ -10,46 +10,142 @@ Then, research the question to expand on your answer. Even if you feel you have
...
+
+    When you run the yarn command in your terminal, several files and directories may be modified or created. The most important of these is the yarn.lock file, which is similar to the package-lock.json file generated by npm. The yarn.lock file records the exact version numbers of all the dependencies that are installed in your project, along with any transitive dependencies. This ensures that the same versions of packages are installed on every machine where the project is run, which helps to avoid compatibility issues. In addition to the yarn.lock file, the node_modules directory may also be modified or created when you run yarn. This directory contains all the packages that are installed in your project, including their dependencies. The node_modules directory can be quite large, especially for projects with many dependencies, so it's typically excluded from version control systems like Git. When you run yarn in your terminal, it will check the yarn.lock file to determine which packages need to be installed or updated, and then it will download and install them in the node_modules directory.
+
+---
+
```
Comment (SunkissedQueen) [inline interview-questions.md:34]: nicely done

File: interview-questions.md [hunk 7]
```diff
@@ -10,46 +10,142 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 Researched answer:
 
+- A React component is a modular, self-contained piece of code that represents a part of the user interface. It encapsulates the HTML, CSS, and JavaScript logic required to render a specific UI element or widget. Each component can be used multiple times in different parts of the application, making it a powerful tool for creating complex UIs. For example, a component might be used to render a button, a form, a menu, or any other element that appears on the screen. React components are designed to be reusable and easy to manage. They can be nested inside other components to create more complex UIs, and can communicate with each other through a system of props and events. This modular architecture allows developers to build large-scale applications that are easier to maintain and update over time. Moreover, React components can be shared across different projects, making it easy to reuse code and create consistent user experiences across different applications.
```
Comment (SunkissedQueen) [inline interview-questions.md:43]: good job proving you understand. I would suggest sharing an example of the structure of a react component

File: interview-questions.md [hunk 8]
```diff
@@ -10,46 +10,142 @@ Then, research the question to expand on your answer. Even if you feel you have
...
+
 4. What is the difference between state values and props in React?
 
 Your answer:
 
+- State is used to manage internal data that can change over time within a component, while props are used to pass down external data from a parent component to a child component.
+
 Researched answer:
 
+- State and props in React are two different ways to manage data that affects the behavior of a component. State represents internal data that can change over time, while props are external data that is passed down from a parent component to its children. To give an example, let's say you're building a to-do list app. The list of items would be stored in the state of the main component, since it's an internal piece of data that can be updated by the user. On the other hand, the title of the app could be passed down to a child component as a prop, since it's a piece of data that doesn't change and is only needed for display purposes. In short, state is used for managing internal data that changes over time, while props are used for passing external data between components. Understanding the difference between the two is essential for building scalable and efficient React applications.
```
Comment (SunkissedQueen) [inline interview-questions.md:55]: solid explanation

File: interview-questions.md [hunk 9]
```diff
@@ -10,46 +10,142 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 5. What is the DOM?
 
 Your answer:
 
+- The Document Object Model (DOM) is a programming interface for web documents that represents the page so that programs can change the document structure, style, and content.
+
 Researched answer:
 
+- The DOM is essentially a tree-like structure that represents the HTML (or XML) document as a collection of nodes and objects, which can be manipulated using JavaScript. The topmost node in the DOM is the document node, which represents the entire HTML document. From there, each element, attribute, and text node is represented by a corresponding node in the DOM tree. The DOM provides a way for web developers to programmatically interact with and modify the content and structure of a web page. This includes tasks such as adding or removing elements, changing the style or layout of elements, and responding to user events such as clicks or mouse movements. In summary, the DOM is an essential part of web development and provides a powerful tool for creating dynamic and interactive web pages using JavaScript.
+
```
Comment (SunkissedQueen) [inline interview-questions.md:68]: good explanation

File: interview-questions.md [hunk 10]
```diff
@@ -10,46 +10,142 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 Your answer:
 
+-  The main difference between the div and span tags is that the div tag is a block-level element used to create larger sections of content on the page, while the span tag is an inline element used to group together small pieces of content within a larger block of text.
+
 Researched answer:
 
+- In HTML, the div and span tags are both used to group content together, but they're used for different purposes. The div tag is used to create a new section or area on the page, like a container that can hold other elements. It's great for separating content into different parts of the page, and you can style it with CSS to control how it looks. The span tag, on the other hand, is used to group together small bits of content within a larger block of text. It's perfect for applying styles or formatting to specific words or phrases within a paragraph or heading. The span tag doesn't create a new line before or after the element, so it only takes up as much space as it needs.
+
+---
+
```
Comment (SunkissedQueen) [inline interview-questions.md:82]: You are doing really well with your explanations. Let's start adding your wonderful personality in here. Next assessment let's create some examples or add your experience using the tool.

File: react-debugging.html [hunk 11]
```diff
@@ -92,3 +106,5 @@ <h1>Draw a Card</h1>
     </script>
   </body>
 </html>
+
+<!--This was extremely fun and I could spend all day going thru code and fixing bugs like this! -->
```
Comment (SunkissedQueen) [inline react-debugging.html:110]: Glad you enjoyed it. Coding bugs extermination is part of the occupation

File: react-debugging.html [hunk 12]
```diff
@@ -65,22 +65,36 @@
...
+
+            {/*  Bug: <button onClick={}>  */}
+            {/*  "JSX attributes must only be assigned a non-empty expression." - In JSX, you cannot assign an attribute to an empty or undefined value. You must provide a non-empty expression, such as a string or a variable, as the attribute value. The missing expression is drawCard.  */}
+
+            <button onClick={drawCard}>
+              Click to Draw a Card
+            </button>
+
+            {/*  Bug: <button onclick={shuffle}>  */}
+            {/*  "Invalid event handler property `onclick`. Did you mean `onClick`? - The warning message is telling us that there is a typo in the button element code. The onclick attribute should be onClick according to React's naming conventions for event handler attributes in JSX. By changing the onclick attribute to onClick, the event handler will be recognized by React and the code will work as intended.  */}
```
Comment (SunkissedQueen) [inline react-debugging.html:83]: Thank you for being so thorough. Really makes you stand out

**Review by SunkissedQueen — APPROVED** [review]
Elmer, you are such a trooper. Thank you for still pouring lots of effort into your assessment. You are approved to merge your changes. You are a more than competent developer; let's get the work to support who you are. This assessment is at the high end of an advanced beginner. Next assessment, let's incorporate some strategies:  
1. On the interview questions. You are doing really well with your explanations. Let's start adding your wonderful personality. You can create some unique examples to portray your thoughts or add your experience using the tool. 
2. On the jest test section of code challenges. In the context of the functions created for the challenges, the test case is only looking at one behavior what the function call will output based on the argument. Next week, try to create one test case for the functionality and add as many expect statements to cover the different edge cases for the function.
```js
describe("", () => {
  it("", () => {
    //test variables
    expect(function call).toEqual(expected output)
    expect(function call).toEqual(expected output)
  })
})
```
Thank you again for all the time and effort you pour into this assessment.
